### PR TITLE
Add admin CLI to change "attach_as" status of the disk.

### LINKED ISSUE #569: Need Admin CLI coverage and unit test coverage for delta disk handling code
#547 introduces proper handling of deltadisks and allows to define a volume  as participating in VM snapshot. However, to complete #390 we still neeed to do the following: 
* Add Admin CLI option to change `attach_as` ("snapshottable") status of the disk (format  TBD)
* Add Admin CLI information about existing deltadisk in the chain (format TBD)
* Add automated test: 
  * attach as default. Check the VM.config to make sure it is attached correctly
  * Change ths status. Make sure VM.config changed. 
  * generate fake .vmdk and -0000x.vmdk files and add unit tests for vmdk_utils handling this 

//CC @pdhamdhere

### PR DESCRIPTION
**Pull request by lipingxue**
Fixes #569.

**Manual testing:**

- create a docker volume with "attach_as=persistent" operation.
```
root@photon-JQQBWNwG6 [ ~ ]# docker volume create --driver=vmdk --name=test4 -o size=100mb -o attach_as=persistent
test4

[root@localhost:~] cat /vmfs/volumes/datastore1/dockvols/test4-47a1d3895cc90e74.vmfd 
{"status": "detached", "volOpts": {"size": "100mb", "attach_as": "persistent"}, "created-by": "photon4", "created": "Wed Aug 10 21:01:06 2016"}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                
```
- create a docker volume with "attach_as=independent_persistent" operation.
```
root@photon-JQQBWNwG6 [ ~ ]# docker volume create --driver=vmdk --name=test5 -o size=100mb -o attach_as=independent_persistent
test5

[root@localhost:~] cat /vmfs/volumes/datastore1/dockvols/test5-
test5-46aedf31be436180.vmfd  test5-flat.vmdk
[root@localhost:~] cat /vmfs/volumes/datastore1/dockvols/test5-46aedf31be436180.vmfd 
{"status": "detached", "volOpts": {"size": "100mb", "attach_as": "independent_persistent"}, "created-by": "photon4", "created": "Wed Aug 10 21:05:03 2016"}   
```

- use "set" command with "operation==attach_as=persistent" to change volume "test5" to "attach_as=persistent"
```
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py  set --volume="/vmfs/volumes//datastore1/dockvols/test5" --options="attach_as=persistent"
{'attach_as': 'persistent'}
Successfully updated settings for : /vmfs/volumes//datastore1/dockvols/test5

[root@localhost:~] cat /vmfs/volumes/datastore1/dockvols/test5-46aedf31be436180.vmfd 
{"status": "detached", "volOpts": {"attach_as": "persistent", "size": "100mb"}, "created-by": "photon4", "created": "Wed Aug 10 21:05:03 2016"} 
```
**Automatic testing:**
automatic testing are added to test:
1. vmdk_create with "attach_as" option.
2. "set" command with "attach_as" option.

make build-esx test passed with those changes.

### CODE REVIEW

**Comment by pdhamdhere** [comment]
@aspear Do you know why vmwclabot is complaining in CI?

**Comment by aspear** [comment]
Hi @pdhamdhere, sorry I don't know.  @travisfinch is the expert, looping him in.

**Comment by travisfinch** [comment]
I will take a look at this as soon as I can. It should be fine to ignore this status check though if you want to merge now.

**Comment by travisfinch** [comment]
If you look at pulls/598, this was submitted by @lipingxue using a different account, which prompted her to sign the CLA (not a VMware organization member).

GitHub statuses are tracked by commit hash, and both PRs have the same commit hash, so the status from the other pull request was pulled in.

As I said in the previous comment, this status is safe to ignore. I will open an issue in the CLA portal repo and fix as soon as I can.

Thanks for looping me in!

File: esx_service/cli/vmdkops_admin_test.py [hunk 1]
```diff
@@ -135,6 +135,7 @@ def test_set_invalid_options(self):
 
     def test_set_invalid_options(self):
         self.assert_parse_error('set --options="acces=read-write"')
+        self.assert_parse_error('set --options="access=persisten"')
```
Comment (govint) [inline esx_service/cli/vmdkops_admin_test.py]: 'attach_as' vs. 'access'

File: esx_service/vmdk_ops.py [hunk 2]
```diff
@@ -1052,25 +1052,36 @@ def set_vol_opts(name, options):
        msg = 'Volume {0} not found.'.format(vol_name)
        logging.warning(msg)
        return False
-       
+   
     # For now only allow resetting the access mode.
```
Comment (govint) [inline esx_service/vmdk_ops.py]: Correct comment to include the new attr.

File: esx_service/cli/vmdkops_admin_test.py [hunk 3]
```diff
@@ -145,6 +146,16 @@ def test_set(self):
...
+        self.assertEqual(args.func, vmdkops_admin.set_vol_opts)
+        self.assertEqual(args.volume, 'vol_name')
+        self.assertEqual(args.options, '"attach_as=persistent"')
+
+        args = self.parser.parse_args('set --volume=vol_name --options="attach_as=independent_persistent"'.split())
+        self.assertEqual(args.func, vmdkops_admin.set_vol_opts)
+        self.assertEqual(args.volume, 'vol_name')
+        self.assertEqual(args.options, '"attach_as=independent_persistent"')
+
     # Usage is always printed on a parse error. It's swallowed to prevent clutter.
```
Comment (govint) [inline esx_service/cli/vmdkops_admin_test.py]: Please add tests to provide the volume name as vol-name@datastore because thats whats accepted.

File: esx_service/vmdk_ops.py [hunk 4]
```diff
@@ -1052,25 +1052,36 @@ def set_vol_opts(name, options):
...
+            logging.warning(msg)
+            return False
+
+    if kv.ATTACH_AS in opts.keys():    
+        if not opts[kv.ATTACH_AS] in kv.ATTACH_AS_TYPES:
+            msg = 'Invalid option value {0}.\n'.format(opts[kv.ATTACH_AS]) +\
+                    'Supported values are {0}.\n'.format(kv.ATTACH_AS_TYPES)
+            logging.warning(msg)
+            return False
+
```
Comment (govint) [inline esx_service/vmdk_ops.py]: This code  block isn't easy to extend if each option needs to repeat the above steps. Instead,

a. change valid opts to a dict of {'<opt>': [valid choices], ....} then

b. loop over the keys in the provided options and see if its present in valid_opts.heys{}. If yes, then
     c. check if the option value is in the list of valid choices for that option. 

     d. If not then
            e. prepare an error message and remember that the option value is invalid.

            f. Exit the loop

g. Figure if there was an error in parsing the options and option values. If yes then log and exit.

Else the above code blocks will get duplicated over and over for each new optioin thats added to the set command.

File: esx_service/vmdk_ops.py [hunk 5]
```diff
@@ -1052,25 +1052,36 @@ def set_vol_opts(name, options):
...
+   
 
     vol_meta = kv.getAll(vmdk_path)
     if vol_meta:
-       vol_meta[kv.VOL_OPTS][kv.ACCESS] = opts[kv.ACCESS]
+       if kv.ACCESS in opts.keys(): 
+           vol_meta[kv.VOL_OPTS][kv.ACCESS] = opts[kv.ACCESS]
+       if kv.ATTACH_AS in opts.keys(): 
+           vol_meta[kv.VOL_OPTS][kv.ATTACH_AS] = opts[kv.ATTACH_AS]    
        return kv.setAll(vmdk_path, vol_meta)
```
Comment (govint) [inline esx_service/vmdk_ops.py]: If vol_meta[kv.VOL_OPTS] wasn't originally there then pls. initialize it to an empty dict first.

File: esx_service/vmdk_ops.py [hunk 6]
```diff
@@ -1052,25 +1052,36 @@ def set_vol_opts(name, options):
...
+            msg = 'Invalid option value {0}.\n'.format(opts[kv.ATTACH_AS]) +\
+                    'Supported values are {0}.\n'.format(kv.ATTACH_AS_TYPES)
+            logging.warning(msg)
+            return False
+   
 
     vol_meta = kv.getAll(vmdk_path)
     if vol_meta:
-       vol_meta[kv.VOL_OPTS][kv.ACCESS] = opts[kv.ACCESS]
+       if kv.ACCESS in opts.keys():
```
Comment (govint) [inline esx_service/vmdk_ops.py]: If vol_meta[kv.VOL_OPTS] wasn't originally there then pls. initialize it to an empty dict first.

File: esx_service/vmdk_ops_test.py [hunk 7]
```diff
@@ -129,6 +133,43 @@ def testAccessOpts(self):
...
+
+        err = vmdk_ops.removeVMDK(self.name)
+        logging.info(err)
+        self.assertEqual(err, None, err)
+    
+    def testAttachAsOpts(self):
+        err = vmdk_ops.createVMDK(vm_name=self.vm_name,
+                                  vmdk_path=self.name,
+                                  vol_name=self.volName,
                                   opts=self.invalid_access_choice)
```
Comment (govint) [inline esx_service/vmdk_ops_test.py]: Are the attach_as attr tests in the wrong function.

File: esx_service/vmdk_ops_test.py [hunk 8]
```diff
@@ -129,6 +133,43 @@ def testAccessOpts(self):
         err = vmdk_ops.createVMDK(vm_name=self.vm_name,
                                   vmdk_path=self.name,
                                   vol_name=self.volName,
+                                  opts=self.invalid_attach_as_choice)
+        logging.info(err)
```
Comment (andrewjstone) [inline esx_service/vmdk_ops_test.py]: I would remove all these logging messages from the test.

**Comment by andrewjstone** [comment]
I agree with all @govint's comments. Those things need to be fixed before this gets merged. Additionally, I think that there should be a way to retrieve this data via the admin cli. Currently the user can set a disk to 'independent_persistent', etc... but can't check what value is set in the admin CLI. That seems wrong to me. The same is true for access permissions. I think we should either add `access type` and `attach type` to the `ls` command or add a new command `ls volume opts` where we just list the options if `ls` is getting to be too big of a table. I don't think we should release this code without that. My worry is that this gets merged then goes into the next release, but the admin visibility support doesn't exist yet. For that reason I think those changes should be part of this PR as well.

**Comment by lipingxue** [comment]
I agree with Andy. We should have some way to show the attach_as and access option after we change it through "set" command. I plan to add this to the existing "ls" command.

**Comment by lipingxue** [comment]
Add "access" and 'attach-as" in "ls" command output.

```
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py ls -h
usage: vmdkops_admin.py ls [-h] [-c Col1,Col2,...]

optional arguments:
  -h, --help        show this help message and exit
  -c Col1,Col2,...  Display selected columns: Choices = ['volume',
                    'datastore', 'created-by', 'created', 'attached-to',
                    'policy', 'capacity', 'used', 'access, attach-as']
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py ls 
Volume  Datastore   Created By VM  Created                   Attached To VM  Policy  Capacity  Used     Access Mode  Attach Mode             
------  ----------  -------------  ------------------------  --------------  ------  --------  -------  -----------  ----------------------  
test1   datastore1  photon3        Mon Aug  8 18:56:37 2016  photon4         N/A     100.00MB  14.00MB  read-write   independent_persistent  
test2   datastore1  photon4        Wed Aug 10 02:14:09 2016  detached        N/A     100.00MB  14.00MB  read-write   persistent              
test3   datastore1  photon4        Wed Aug 10 12:10:26 2016  detached        N/A     100.00MB  14.00MB  read-write   independent_persistent  
test4   datastore1  photon4        Wed Aug 10 21:01:06 2016  detached        N/A     100.00MB  14.00MB  read-write   persistent              
test5   datastore1  photon4        Wed Aug 10 21:05:03 2016  detached        N/A     100.00MB  14.00MB  read-write   independent_persistent  
```

use "set" command to change "access=read-only" and "attach-as=persistent" for volume "test5"
```
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py set --volume=test5@datastore1 --options="attach_as=persistent, access=read-only"
Successfully updated settings for : test5@datastore1

[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py ls
Volume  Datastore   Created By VM  Created                   Attached To VM  Policy  Capacity  Used     Access Mode  Attach Mode             
------  ----------  -------------  ------------------------  --------------  ------  --------  -------  -----------  ----------------------  
test1   datastore1  photon3        Mon Aug  8 18:56:37 2016  photon4         N/A     100.00MB  14.00MB  read-write   independent_persistent  
test2   datastore1  photon4        Wed Aug 10 02:14:09 2016  detached        N/A     100.00MB  14.00MB  read-write   persistent              
test3   datastore1  photon4        Wed Aug 10 12:10:26 2016  detached        N/A     100.00MB  14.00MB  read-write   independent_persistent  
test4   datastore1  photon4        Wed Aug 10 21:01:06 2016  detached        N/A     100.00MB  14.00MB  read-write   persistent              
test5   datastore1  photon4        Wed Aug 10 21:05:03 2016  detached        N/A     100.00MB  14.00MB  read-only    persistent    

[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py ls -c volume,access,attach-as
Volume  Access Mode  Attach Mode             
------  -----------  ----------------------  
test1   read-write   independent_persistent  
test2   read-write   persistent              
test3   read-write   independent_persistent  
test4   read-write   persistent              
test5   read-only    persistent      
```

**Comment by lipingxue** [comment]
I have addressed all comments and repush the change. Please review it.

File: esx_service/cli/vmdkops_admin.py [hunk 9]
```diff
@@ -388,8 +389,11 @@ def generate_ls_rows():
         policy = get_policy(metadata, path)
         size_info = get_vmdk_size_info(path)
         created, created_by = get_creation_info(metadata)
+        access = get_access(metadata)
```
Comment (msterin) [inline esx_service/cli/vmdkops_admin.py]: i think the names should be get_access_mode() and get_attach_mode() ? this way it will consistent with the column names , now it's a bit confusing.

File: esx_service/cli/vmdkops_admin_sanity_test.py [hunk 10]
```diff
@@ -37,7 +37,7 @@ def test_ls(self):
         output = subprocess.check_output(cmd, shell=True, stderr=self.devnull)
         lines = output.split('\n')
         divider_columns = lines[1].split()
-        self.assertEqual(8, len(divider_columns))
+        self.assertEqual(10, len(divider_columns))
```
Comment (msterin) [inline esx_service/cli/vmdkops_admin_sanity_test.py]: this used to be (and still is) hard to read. Can you add a variable so there is a name involved? Something like 
`expected_column_count = 10`

File: esx_service/cli/vmdkops_admin_test.py [hunk 11]
```diff
@@ -135,16 +135,27 @@ def test_set_invalid_options(self):
...
+        args = self.parser.parse_args('set --volume=vol_name@datastore --options="attach-as=persistent"'.split())
+        self.assertEqual(args.func, vmdkops_admin.set_vol_opts)
+        self.assertEqual(args.volume, 'vol_name@datastore')
+        self.assertEqual(args.options, '"attach-as=persistent"')
+
+        args = self.parser.parse_args('set --volume=vol_name@datastore --options="attach-as=independent_persistent"'.split())
+        self.assertEqual(args.func, vmdkops_admin.set_vol_opts)
+        self.assertEqual(args.volume, 'vol_name@datastore')
+        self.assertEqual(args.options, '"attach-as=independent_persistent"')
+
```
Comment (msterin) [inline esx_service/cli/vmdkops_admin_test.py]: good stuff ! 
Please do add a few negative tests (wrong attach mode string) to make sure it is handled properly.
Also, a test for unicode or non-English char set passed to attach-as needs to be added.

File: esx_service/cli/vmdkops_admin_test.py [hunk 12]
```diff
@@ -211,7 +222,7 @@ def test_ls_no_args(self):
         volumes = vmdk_utils.get_volumes()
         header = vmdkops_admin.all_ls_headers()
         rows = vmdkops_admin.generate_ls_rows()
-        self.assertEqual(8, len(header))
+        self.assertEqual(10, len(header))
```
Comment (msterin) [inline esx_service/cli/vmdkops_admin_test.py]: same comment as above - a var with a name, instead of '10', would make it easier to read

File: esx_service/vmdk_ops_test.py [hunk 13]
```diff
@@ -81,6 +81,10 @@ class VmdkCreateRemoveTestCase(unittest.TestCase):
     invalid_access_choice = {volume_kv.ACCESS: u'only-read'}
     invalid_access_opt = {u'acess': u'read-write'}
     valid_access_opt = {volume_kv.ACCESS: 'read-only'}
+    invalid_attach_as_choice = {volume_kv.ATTACH_AS: u'persisten'}
```
Comment (msterin) [inline esx_service/vmdk_ops_test.py]: nice !

**Comment by lipingxue** [comment]
Address comments from Mark, and repush the change.

**Comment by msterin** [comment]
@lipingxue - please address my other comments (negative test in CLI, function rename) - either change the code, or reply in the comments if you disagree

File: esx_service/cli/vmdkops_admin.py [hunk 14]
```diff
@@ -118,7 +118,8 @@ def commands():
                 '-c': {
                     'help': 'Display selected columns',
                     'choices': ['volume', 'datastore', 'created-by', 'created',
-                                'attached-to', 'policy', 'capacity', 'used'],
+                                'attached-to', 'policy', 'capacity', 'used',
+                                'access', 'attach-as'],
```
Comment (govint) [inline esx_service/cli/vmdkops_admin.py]: May be rename the columns as "access-mode" and "attach-mode". Plus documentation on what these choices are    in "vmdkops-admin-cli-spec.md" mentioning what these are,

File: esx_service/vmdk_ops.py [hunk 15]
```diff
@@ -1052,25 +1052,37 @@ def set_vol_opts(name, options):
        msg = 'Volume {0} not found.'.format(vol_name)
        logging.warning(msg)
        return False
-       
-    # For now only allow resetting the access mode.
-    valid_opts = [kv.ACCESS]
-    invalid = frozenset(opts.keys()).difference(valid_opts)
+   
+    # For now only allow resetting the access and attach_as mode.
+    valid_opts = [kv.ACCESS, kv.ATTACH_AS]
```
Comment (govint) [inline esx_service/vmdk_ops.py]: can delete this line

File: esx_service/vmdk_ops.py [hunk 16]
```diff
@@ -1052,25 +1052,37 @@ def set_vol_opts(name, options):
...
+                has_invalid_opt_value = True
+                
+    if has_invalid_opt_value:
+        return False   
+    
     vol_meta = kv.getAll(vmdk_path)
     if vol_meta:
-       vol_meta[kv.VOL_OPTS][kv.ACCESS] = opts[kv.ACCESS]
+       if not vol_meta[kv.VOL_OPTS]:
+           vol_mata[kv.VOL_OPTS] = []
```
Comment (govint) [inline esx_service/vmdk_ops.py]: vol_m"e"ta  :)

File: esx_service/vmdk_ops_test.py [hunk 17]
```diff
@@ -118,39 +122,64 @@ def testBadOpts(self):
...
                                   vol_name=self.volName,
                                   opts=self.badOpts)
-        logging.info(err)
         self.assertNotEqual(err, None, err)
 
         err = vmdk_ops.removeVMDK(self.name)
-        logging.info(err)
         self.assertNotEqual(err, None, err)
-
+
```
Comment (govint) [inline esx_service/vmdk_ops_test.py]: Nit, white space added

**Comment by lipingxue** [comment]
Hi Mark, @msterin 
I am not sure what are the negative tests you are mentioned. I checked the code in vmdkops_admin_test.py, the place you asked me to add negative test is inside "class TestParsing". 
Based on my understanding, all the tests inside this class just to test whether the command line can be parsed to args correctly or not. It does not test any command functionality here. I am not sure what is the negative test we should put here. 

In this file, the functionality test for command "set" is missing (even for existing option "--option=access"). I think we should add functionality test for command "set", both positive and negative test.  What do you think?

**Comment by msterin** [comment]
Agreed -  we need test for set.

By "negative" test I meant test that are expected to fail. specifically  self.parser.parse_args with (some) wrong options is expected to fail; for arg parse this is low priority though so feel free to skip.

File: esx_service/cli/vmdkops_admin.py (hunk 14 above)
Comment (lipingxue) [inline esx_service/cli/vmdkops_admin.py]: I plan to change the column name in CLI output to "access" and "attach-as". Otherwise, it will need a lot of changes in the code. There is no extra advantage to name those two columns as "access-mode" and "attach-mode".

File: esx_service/cli/vmdkops_admin.py (hunk 9 above)
Comment (lipingxue) [inline esx_service/cli/vmdkops_admin.py]: I plan to change the column names back to "access" and "attach-as". So keep the change as it is to make them consistent.

File: esx_service/vmdk_ops.py (hunk 15 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: nice catch :)

File: esx_service/vmdk_ops.py (hunk 16 above)
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: nice catch, a typo.

**Comment by lipingxue** [comment]
Repush the change.
Change the column name and options to "access" and "attach-as".
Add code to do functionality testing for "set" command
Address minor comments from Govindan.

I am trying to setup an environment to test the change using vsan datastore.

File: esx_service/cli/vmdkops_admin_test.py [hunk 18]
```diff
@@ -228,6 +229,85 @@ def test_ls_no_args(self):
...
+        return [x
+                for x in vmdk_utils.get_volumes()
+                if x['filename'].startswith('testvol')]
+
+    def test_set_access(self):
+        volumes = self.get_testvols()
+        self.assertEqual(len(volumes), self.vol_count)
+        for v in volumes:
+            attach_as_opt = random.choice(kv.ATTACH_AS_TYPES)
+            vol_arg = '@'.join([v['filename'].replace('.vmdk', ''), v['datastore']])
```
Comment (msterin) [inline esx_service/cli/vmdkops_admin_test.py]: nit: please add a comment about intended result of this operation (i.e. "generates string like '...' ") , it is hard to parse if the line is doing what it intents to do s the intent is not clear.

**Comment by lipingxue** [comment]
I think the change I made is agnostic to underline datastore. So I plan to merge after the CI test pass.
